model.py

```diff
@@ -6830,6 +6836,136 @@ def update_credentials(self, _db, access_token, credential):
+        self._client_secret = unicode(bcrypt.hashpw(
+            plaintext_secret, bcrypt.gensalt()
+        ))
+
+    def _correct_secret(self, plaintext_secret):
+        """Determines if a plaintext string is the client_secret"""
+        return (bcrypt.hashpw(plaintext_secret, self.client_secret)
+                == self.client_secret)
+
+    @classmethod
+    def register(cls, _db, name, is_source=True):
+        """Creates a new collection with client details and a datasource."""
+
+        name = unicode(name)
+        collection = get_one(_db, cls, name=name)
+        if collection:
+            raise ValueError(
+                "A collection with the name '%s' already exists: %r" % (
+                name, collection)
+            )
+
+        collection_data_source = None
+        if is_source:
+            collection_data_source, ignore = get_one_or_create(
+                _db, DataSource, name=name, offers_licenses=False
+            )
+
+        client_id, plaintext_client_secret = cls.generate_client_details()
+        # Generate a new client_id if it's not unique initially.
+        while get_one(_db, cls, client_id=client_id):
+            client_id, plaintext_client_secret = cls.generate_client_details()
+
+        kw = dict(
+            client_id=unicode(client_id),
+            client_secret=unicode(plaintext_client_secret)
+        )
+
+        if collection_data_source:
+            kw['data_source'] = collection_data_source
+
+        collection, ignore = get_one_or_create(_db, cls, name=name, **kw)
+        _db.commit()
+
+        return collection, plaintext_client_secret
+
+    @classmethod
+    def generate_client_details(cls):
+        client_id_chars = ('abcdefghijklmnopqrstuvwxyz'
+                           'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+                           '0123456789')
+        client_secret_chars = client_id_chars + '!"#$%&()*+,-./[]^_`{}|~'
+
+        def make_client_string(chars, length):
+            return u"".join([random.choice(chars) for x in range(length)])
+        client_id = make_client_string(client_id_chars, 25)
+        client_secret = make_client_string(client_secret_chars, 40)
+
+        return client_id, client_secret
+
+    @classmethod
+    def authenticate(cls, _db, client_id, plaintext_client_secret):
+        collection = get_one(_db, cls, client_id=client_id)
+        if (collection and
+            collection._correct_secret(plaintext_client_secret)):
+            return collection
+        return None
+
+    def catalog_identifier(self, _db, identifier):
+        """Creates a LicensePool to catalog an identifier for a collection"""
+
+        license_pool = identifier.licensed_through
+        if not license_pool:
+            license_pool, new = get_one_or_create(
+                _db, LicensePool, identifier=identifier
+            )
+        if license_pool not in self.license_pools:
+            self.license_pools.append(license_pool)
+
+    def catalog(self, _db):
```

Comment:
something like this might be a more efficient query for the identifiers:
```
qu = _db.query(Identifier).join(LicensePools).join(Collection).filter(Collection.id=self.id)
```